File: python/xrun_tui/src/xrun_tui/screens/help.py
from __future__ import annotations

from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Vertical, VerticalScroll
from textual.screen import ModalScreen
from textual.widgets import Static

# Section → list of (keys, description)
_HELP: list[tuple[str, list[tuple[str, str]]]] = [
    ("Global", [
        ("?",          "This help"),
        ("ctrl+p",     "Command palette"),
        ("g d",        "Dashboard"),
        ("g r",        "Runs"),
        ("g i",        "Instances"),
        ("g v",        "Vendors"),
        ("g h",        "Doctor (health)"),
        ("g l",        "Launch"),
        (",",          "Settings"),
        ("ctrl+r / F5","Refresh current screen"),
        ("q / ctrl+c", "Quit"),
    ]),
    ("Lists & tables", [
        ("j / k or ↑ ↓", "Move cursor"),
        ("enter",         "Open detail"),
        ("f or /",        "Fuzzy filter"),
    ]),
    ("Runs", [
        ("s",      "Stop selected run"),
        ("r",      "Rerun selected run"),
        ("p",      "Pull latest artifacts"),
        ("c",      "Toggle compare selection"),
        ("C",      "Open compare (2 selected)"),
        ("G",      "Toggle grouping by manifest"),
        ("e",      "Export visible runs (JSON)"),
        ("f or /", "Fuzzy filter"),
    ]),
    ("Run detail", [
        ("1 / 2 / 3 / 4", "Stages / Logs / Manifest / Metrics"),
        ("a",              "Artifacts browser"),
        ("Relaunch btn",   "Launch from same manifest"),
        ("esc",            "Back"),
    ]),
    ("Instances", [
        ("x",      "Destroy selected remote instance"),
        ("ctrl+r", "Refresh"),
    ]),
    ("Forms", [
        ("ctrl+s", "Save"),
        ("ctrl+t", "Test connection"),
        ("esc",    "Cancel / back"),
    ]),
]


class HelpScreen(ModalScreen[None]):
    BINDINGS = [
        Binding("escape,q,?", "dismiss", show=False),
    ]

    DEFAULT_CSS = """
    HelpScreen { align: center middle; }
    #help-box {
        background: #24283b;
        border: round #7aa2f7;
        width: 70;
        height: 32;
        padding: 1 2;
    }
    #help-title {
        color: #7aa2f7;
        text-style: bold;
        height: 1;
        padding-bottom: 1;
        border-bottom: solid #414868;
    }
    .help-section-title {
        color: #bb9af7;
        text-style: bold;
        padding: 1 0 0 0;
        height: auto;
    }
    .help-row { height: 1; padding-left: 2; }
    #help-footer {
        color: #565f89;
        height: 1;
        padding-top: 1;
        border-top: solid #414868;
        content-align: center middle;
    }
    """

    def compose(self) -> ComposeResult:
        with Vertical(id="help-box"):
            yield Static("xrun  ·  Keyboard reference", id="help-title")
            with VerticalScroll():
                for section, rows in _HELP:
                    yield Static(section, classes="help-section-title")
                    for keys, desc in rows:
                        yield Static(
                            f"  [bold #7dcfff]{keys:<14}[/] [#c0caf5]{desc}[/]",
                            classes="help-row",
                        )
            yield Static("press [#7aa2f7]?[/] or [#7aa2f7]esc[/] to close",
                        id="help-footer")

    def action_dismiss(self) -> None:
        self.dismiss(None)

File: crates/xrun-tui/src/view/help.rs
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders, Clear, Paragraph, Wrap};

use crate::state::{AppState, Screen};

pub fn render(f: &mut Frame, state: &AppState) {
    let area = super::centered_rect(70, 80, f.area());
    f.render_widget(Clear, area);
    let block = Block::default()
        .title(" Help \u{2014} any key to close ")
        .borders(Borders::ALL)
        .border_style(state.theme.border);

    let screen_help = match &state.screen {
        Screen::Runs => concat!(
            "Runs screen\n",
            "\n",
            "  enter       open run detail\n",
            "  L           launch picker\n",
            "  I           instances\n",
            "  ,           settings\n",
            "  S           stop run (active)\n",
            "  P           pull checkpoint (recent)\n",
            "  R           rerun\n",
            "  /           filter by name\n",
            "  Tab         switch active / recent\n",
            "  g g / G     top / bottom\n",
        ),
        Screen::RunDetail(_, _) => concat!(
            "Run detail\n",
            "\n",
            "  Tab / Shift-Tab   switch tabs\n",
            "  q / Esc           back\n",
            "\n",
            "  Logs tab\n",
            "  Space             toggle autoscroll\n",
            "  k / j             scroll up / down\n",
            "  Home / End        top / bottom\n",
            "\n",
            "  Manifest tab\n",
            "  e                 open in $EDITOR\n",
        ),
        Screen::Launch => concat!(
            "Launch picker\n",
            "\n",
            "  k / j     navigate up / down\n",
            "  enter     launch selected\n",
            "  Esc       back\n",
            "  g g / G   top / bottom\n",
        ),
        Screen::Instances => concat!(
            "Instances\n",
            "\n",
            "  Tab       toggle Local / Remote (vast.ai)\n",
            "  k / j     navigate up / down\n",
            "  D         destroy orphan (local only)\n",
            "  Esc       back\n",
            "  g g / G   top / bottom\n",
        ),
        Screen::Settings => concat!(
            "Settings\n",
            "\n",
            "  k / j     navigate up / down\n",
            "  enter     edit selected\n",
            "  Esc       back / cancel edit\n",
            "  g g / G   top / bottom\n",
        ),
        Screen::Vendors => concat!(
            "Vendors\n",
            "\n",
            "  k / j     navigate up / down\n",
            "  e/enter   edit credentials (masked)\n",
            "  i         import native key file\n",
            "  t         test connection\n",
            "  r         revoke credentials\n",
            "  Esc / q   back\n",
        ),
    };

    let global_help = concat!(
        "Global\n",
        "\n",
        "  q / Esc   quit / back\n",
        "  ?         this help\n",
        "  :         command palette\n",
        "  g g / G   top / bottom (list screens)\n",
        "\n",
    );

    let full_help = format!("{}{}", global_help, screen_help);
    f.render_widget(
        Paragraph::new(full_help)
            .block(block)
            .wrap(Wrap { trim: false }),
        area,
    );
}
